### PR TITLE
lv2: Store and make use of the data passed to sys_spu_initialize()

### PR DESCRIPTION
**Pull request by effective-light**

### CODE REVIEW

**Comment by elad335** [comment]
No, if it was so simple I would have implemented it already.

**Comment by effective-light** [comment]
@elad335 In that case, do you know if there is any documentation regarding the issue?

**Comment by elad335** [comment]
sys_spu_initialize needs to check `current amount of non-scheduled grouped SPU threads + largest scheduled SPU group thread's count < max_usable_spu`, also check `raw spu count + non-scheduled spu threads count < 5`, also check `raw spu threads count < max_usable_raw_spu`. If any are false EAGAIN is returned and it all must be thread safe with group/raw spu creation.

File: rpcs3/Emu/Cell/lv2/sys_spu.cpp [hunk 1]
```diff
@@ -505,7 +508,7 @@ error_code sys_spu_thread_group_create(ppu_thread& ppu, vm::ptr<u32> id, u32 num
 	bool use_scheduler = true;
 	bool use_memct = !!(type & SYS_SPU_THREAD_GROUP_TYPE_MEMORY_FROM_CONTAINER);
 	bool needs_root = false;
-	u32 max_threads = 6; // TODO: max num value should be affected by sys_spu_initialize() settings
+	u32 max_threads = ppu.max_usable_spu;
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_spu.cpp:511]: This value must remain 6, what you want is an EAGAIN check afterwards.

File: rpcs3/Emu/Cell/PPUThread.h [hunk 2]
```diff
@@ -297,6 +297,9 @@ class ppu_thread : public cpu_thread
 
 	u32 dbg_step_pc = 0;
 
+	u32 max_usable_spu = 6;
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUThread.h:300]: This is thread-local context, you need to use g_fxo container.

**Review by elad335 — CHANGES_REQUESTED** [review]

**Comment by elad335** [comment]
I forgot, in sys_spu_thread_group_create you need to recheck the first check of sys_spu_initiaize I mentioned with the inrormation of tge group to be created and return EAGAIN if false.

**Comment by elad335** [comment]
Here's the explenation behind this check: non-scheduled grouped SPU threads always take up hw SPU threads, while with scheduled the OS can supend the execution of SPU group if it currently doesn't have enough hw threads. But there must be room for atleast one group of this type to run and it can be any if them, hence why you need to get know what is the size of the klrgest group of this type.

**Pull request closed without merging**